### PR TITLE
[codegen/go] Add support for ResourceType and isComponent

### LINKED ISSUE #5360: Generate component resource SDKs in Go.
Depends on #5357.

### LINKED ISSUE #5412: [codegen/go] Handle ResourceType
Depends on #5325

### PR DESCRIPTION
**Pull request by lblackstone**
Fix #5412
Fix #5360

### CODE REVIEW

File: pkg/codegen/internal/test/testdata/simple-resource-schema/otherResource.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+// NewOtherResource registers a new resource with the given unique name, arguments, and options.
+func NewOtherResource(ctx *pulumi.Context,
+	name string, args *OtherResourceArgs, opts ...pulumi.ResourceOption) (*OtherResource, error) {
+	if args == nil {
+		args = &OtherResourceArgs{}
+	}
+	var resource OtherResource
+	err := ctx.RegisterResource("example::OtherResource", name, args, &resource, opts...)
```
Comment (justinvp) [inline pkg/codegen/internal/test/testdata/simple-resource-schema/otherResource.go:25]: Just a note: In the draft PR of the SDK changes, I've added `RegisterRemoteComponentResource` that should be used for remote components (unless we want to expose this in some other way).

https://github.com/pulumi/pulumi/pull/5558/files#diff-6ac1847413cb55c8d480423da6e23ed6ac98c0066fcf7c57f8b5cd960a04dee1R526

File: pkg/codegen/internal/test/testdata/simple-resource-schema/otherResource.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+package example
+
+import (
+	"reflect"
+
+	"github.com/pulumi/pulumi/sdk/v2/go/pulumi"
+)
+
+type OtherResource struct {
+	pulumi.CustomResourceState
```
Comment (justinvp) [inline pkg/codegen/internal/test/testdata/simple-resource-schema/otherResource.go:13]: If it's a component, I believe this should be `pulumi.ResourceState`.

```suggestion
	pulumi.ResourceState
```

File: pkg/codegen/internal/test/testdata/simple-resource-schema/otherResource.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	Foo *Resource `pulumi:"foo"`
+}
+
+type OtherResourceState struct {
+	Foo pulumi.Resource
+}
+
+func (OtherResourceState) ElementType() reflect.Type {
+	return reflect.TypeOf((*otherResourceState)(nil)).Elem()
+}
```
Comment (justinvp) [inline pkg/codegen/internal/test/testdata/simple-resource-schema/otherResource.go:55]: (I'm sure you're already planning to strip these, but I'll mention it anyway). The `Get<resource>` and associated `<resource>State` structs shouldn't be emitted for component resources.

**Comment by lblackstone** [comment]
@pgavlin I made the updates we discussed last week. Can you check to make sure these changes look like you expected?

I also need to update the Construct registration depending on your answer to https://github.com/pulumi/pulumi/pull/5497#discussion_r504301470

**Comment by lblackstone** [comment]
Hmm, probably don't need to generate Input/Output types for Providers, but not sure if it matters or not: https://github.com/pulumi-bot/pulumi-azuread/compare/3fc962f..6320649#diff-42f2dd74963d91d730aa5d29e2e141a6213917686959f08d2dc8cc249348723aR91-R128

File: pkg/codegen/go/gen.go [hunk 4]
```diff
@@ -835,6 +876,24 @@ func (pkg *pkgContext) genResource(w io.Writer, r *schema.Resource) error {
 	fmt.Fprintf(w, "\treturn reflect.TypeOf((*%sArgs)(nil)).Elem()\n", camel(name))
 	fmt.Fprintf(w, "}\n\n")
 
+	// Emit the input types.
```
Comment (justinvp) [inline pkg/codegen/go/gen.go:879]: Nit: I was initially confused what this was for until I read through the rest of the changes. This is the input type for the resource itself? Maybe clarify the comment?

```suggestion
	// Emit the resource input type.
```

File: pkg/codegen/go/gen.go [hunk 5]
```diff
@@ -835,6 +876,24 @@ func (pkg *pkgContext) genResource(w io.Writer, r *schema.Resource) error {
...
+	fmt.Fprintf(w, "type %sInput interface {\n", name)
+	fmt.Fprintf(w, "\tpulumi.Input\n\n")
+	fmt.Fprintf(w, "\tTo%[1]sOutput() %[1]sOutput\n", name)
+	fmt.Fprintf(w, "\tTo%[1]sOutputWithContext(ctx context.Context) %[1]sOutput\n", name)
+	fmt.Fprintf(w, "}\n\n")
+	genInputMethods(w, name, name, name, false)
+
+	// Emit the output types.
```
Comment (justinvp) [inline pkg/codegen/go/gen.go:887]: Nit: Similarly:

```suggestion
	// Emit the resource output type.
```

**Review by justinvp — APPROVED** [review]
LGTM, but would like @pgavlin to review as well.

File: pkg/codegen/go/gen.go (hunk 4 above)
Comment (lblackstone) [reply pkg/codegen/go/gen.go:879]: Yes, each resource now gets a top level Input/Output type. I'll clarify the comment.

**Comment by lblackstone** [comment]
@pgavlin This is RFR

**Pull request merged**